### ISSUE TITLE
gateway/bastion host (feature request)

### ISSUE THREAD

**Issue by jalberto** [kontena/kontena#1491]
For a more secure topology is recommended to use the bastion/fw/gw pattern, where one host controls incoming & outcoming internet access for the other hosts using a combination of routes, NAT & forwading.
This topology avoids the use of public IPs in the nodes.

Right now to use this topology with kontena each host needs to be pre-configured by hand with routes, and fw rules.

The suggestion would be to add ability to master (or a new node type) to work as bastion/gw/fw of a grid.

Advantages:

- more security:
  - is simpler to securize one host than every host
  - no more scans or brutoforce to each server with a public ip
- more efficent:
  - don't need to pay for exra public IPs
  - help the public IP shortage

I can think in several difficulties:

- each provider may require a different strategy (this can be implemented in provider-plugins)
- baremetal may require quite an specific configuration (so instead of provider API kontena must run commands tosetup everything)

**Comment by SpComb** [comment]
Having a single node providing incoming/outgoing network connectivity for a grid doesn't sound like a good idea in terms of high-availability. Using NAT generally forms a bit of a SPOF, and achieving load-balancing/failover is complicated a lot by various low-level networking details that can vary greatly between different cloud providers.

IMO this sounds like something that needs to be dealt with at the provider level... AFAIK this setup describes AWS VPCs pretty closely?

OTOH doing some kind of network routing tricks with containers might provide some interesting opportunities.. think of a Docker container acting as the gateway router for the weave overlay network...

**Comment by SpComb** [comment]
You can take a look at the Microsoft Ananta paper to get an idea about the kinds of complexities involved in cloud-scale NAT: http://dl.acm.org/citation.cfm?id=2486026 [[PDF]](https://www.eecs.umich.edu/courses/eecs589/papers/ananta.pdf)

**Comment by jalberto** [comment]
@SpComb you are right in terms of complexity. But usually that complexity curve si lower as more servers you have (if planned properly of course).
I think this topology doesn't exclude the posibility of open additinal enrypoints for specialized balancers, so both of best worlds can work together (a bastion protecting internal servers + an aditional entry point for an specific service).

In other hand, I am thinking in a lot of small servers running togehter (think ARM & other low powered servers) In a 10 servers network maybe not big deal to have 10 public IPs and one FW in each server, but in a 100-1000 servers network that can become really expensive and complex, is a nightmare to manage fail2ban, fw, & friends in each one.

I started to read the paper, thanks! quite interesting so far. Actually I found Azure the simples provider to use the bastion topology, as the routing & netowrkign is on the azure layer.
